Prometheus exporter causes panic if result of e.deltaCalc.Calc(metricHash, value) is < 0

**Bug description**

Prometheus exporter causes panic if result of  e.deltaCalc.Calc(metricHash, value) is < 0 

https://github.com/openconfig/gnmi-gateway/blob/5dc6155c695ba9e6091db5222ecb52a239d11abd/gateway/exporters/prometheus/prometheus.go#L101

**Step to reproduce**

1. Setup the gnmi-prometheus example as described in https://github.com/openconfig/gnmi-gateway/blob/release/examples/gnmi-prometheus/README.md

2. reload device being polled 

**Expected behavior**
The Prometheus exporter should handle counter resets without a panic

**Output or code snippets**


```
"level":"info","time":"2021-01-17T23:09:08Z","message":"Target lab1-fab1-pod1-leaf1.hyposcaler.net: Connected"}
panic: counter cannot decrease in value

goroutine 49 [running]:
github.com/prometheus/client_golang/prometheus.(*counter).Add(0xc0010d3560, 0xc115413c00000000)
        /go/pkg/mod/github.com/prometheus/client_golang@v1.4.1/prometheus/counter.go:109 +0x116
github.com/openconfig/gnmi-gateway/gateway/exporters/prometheus.(*PrometheusExporter).Export(0xc0005b2270, 0xc004a4e5d0)
        /opt/gnmi-gateway/gateway/exporters/prometheus/prometheus.go:103 +0x22f
github.com/openconfig/gnmi-gateway/gateway.(*CacheClient).run(0xc00478bb00)
        /opt/gnmi-gateway/gateway/gateway.go:150 +0x55
created by github.com/openconfig/gnmi-gateway/gateway.NewCacheClient
        /opt/gnmi-gateway/gateway/gateway.go:136 +0x1bf
 
```

**Environment details**
Please include the output of `go version`. If you have any additional specific details about your environment to share please include them in the bug report.

go version go1.15.5 linux/amd64

Target device being polled was an arista vEOS (ami-036a5d80077d33df2)



**Additional context**

I was able to eliminate the issue by replacing:

https://github.com/openconfig/gnmi-gateway/blob/5dc6155c695ba9e6091db5222ecb52a239d11abd/gateway/exporters/prometheus/prometheus.go#L102-L104

with

```go
  if exists && delta >= 0 {
    m.Add(delta)
  }
```

--- a/gateway/exporters/prometheus/prometheus.go
+++ b/gateway/exporters/prometheus/prometheus.go
@@ -99,7 +99,7 @@ func (e *PrometheusExporter) Export(leaf *ctree.Leaf) {
 		switch m := metric.(type) {
 		case prom.Counter:
 			delta, exists := e.deltaCalc.Calc(metricHash, value)
-			if exists {
+			if exists && delta >= 0 {
 				m.Add(delta)
 			}
 		case prom.Gauge:

--- a/gateway/exporters/prometheus/prometheus_test.go
+++ b/gateway/exporters/prometheus/prometheus_test.go
@@ -13,9 +13,14 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-package prometheus_test
+package prometheus
 
 import (
+	"github.com/openconfig/gnmi-gateway/gateway/configuration"
+	"github.com/openconfig/gnmi/cache"
+	"github.com/openconfig/gnmi/ctree"
+	pb "github.com/openconfig/gnmi/proto/gnmi"
+	"github.com/prometheus/client_golang/prometheus/promauto"
 	"math/rand"
 	"strconv"
 	"testing"
@@ -24,10 +29,9 @@ import (
 	"github.com/stretchr/testify/assert"
 
 	"github.com/openconfig/gnmi-gateway/gateway/exporters"
-	"github.com/openconfig/gnmi-gateway/gateway/exporters/prometheus"
 )
 
-var _ exporters.Exporter = new(prometheus.PrometheusExporter)
+var _ exporters.Exporter = new(PrometheusExporter)
 
 func makeExampleLabels(seed int) prom.Labels {
 	rand.Seed(int64(seed))
@@ -45,8 +49,46 @@ func TestMapHash(t *testing.T) {
 
 	testLabels := makeExampleLabels(2906) // randomly selected consistent seed
 
-	firstHash := prometheus.NewStringMapHash("test_metric", testLabels)
+	firstHash := NewStringMapHash("test_metric", testLabels)
 	for i := 0; i < 100; i++ {
-		assertions.Equal(firstHash, prometheus.NewStringMapHash("test_metric", testLabels), "All hashes of the testLabels should be the same.")
+		assertions.Equal(firstHash, NewStringMapHash("test_metric", testLabels), "All hashes of the testLabels should be the same.")
 	}
+}
+
+func TestPrometheusExporter_Export(t *testing.T) {
+	n := &pb.Notification{
+		Prefix: &pb.Path{Target: "a", Origin: "b"},
+		Update: []*pb.Update{
+			{
+				Path: &pb.Path{
+					Elem: []*pb.PathElem{{Name: "c"}},
+				},
+				Val: &pb.TypedValue{Value: &pb.TypedValue_IntVal{IntVal: -1}},
+			},
+		},
+	}
+
+	metricName, labels := UpdateToMetricNameAndLabels(n.GetPrefix(), n.Update[0])
+	metricHash := NewStringMapHash(metricName, labels)
+
+	// Prime the delta calculator
+	calc := NewDeltaCalculator()
+	calc.Calc(metricHash, 20)
+
+	e := &PrometheusExporter{
+		config:    &configuration.GatewayConfig{},
+		cache:     cache.New(nil),
+		deltaCalc: calc,
+		metrics: map[Hash]prom.Metric{
+			metricHash: promauto.NewCounter(prom.CounterOpts{
+				Name:        metricName,
+				ConstLabels: labels,
+			}),
+		},
+		typeLookup: nil,
+	}
+	assert.NotPanics(t, func() {
+		e.Export(ctree.DetachedLeaf(n))
+	})
+
 }

